modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportStartReindexJobAction.java

```diff
@@ -0,0 +1,195 @@
+    extends TransportMasterNodeAction<StartReindexJobAction.Request, StartReindexJobAction.Response> {
+
+    private final PersistentTasksService persistentTasksService;
+    private final Client client;
+
+    @Inject
+    public TransportStartReindexJobAction(TransportService transportService, ThreadPool threadPool,
+                                          ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                                          ClusterService clusterService, PersistentTasksService persistentTasksService, Client client) {
+        super(StartReindexJobAction.NAME, transportService, clusterService, threadPool, actionFilters, StartReindexJobAction.Request::new,
+            indexNameExpressionResolver);
+        this.persistentTasksService = persistentTasksService;
+        this.client = client;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected StartReindexJobAction.Response newResponse() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected StartReindexJobAction.Response read(StreamInput in) throws IOException {
+        return new StartReindexJobAction.Response(in);
+    }
+
+    @Override
+    protected void masterOperation(Task task, StartReindexJobAction.Request request, ClusterState state,
+                                   ActionListener<StartReindexJobAction.Response> listener) {
+        String generatedId = UUIDs.randomBase64UUID();
+
+        ReindexJob job = new ReindexJob(request.getReindexRequest(), threadPool.getThreadContext().getHeaders());
+
+        // TODO: Task name
+        persistentTasksService.sendStartRequest(generatedId, ReindexTask.NAME, job, new ActionListener<>() {
+            @Override
+            public void onResponse(PersistentTasksCustomMetaData.PersistentTask<ReindexJob> persistentTask) {
+                if (request.getWaitForCompletion()) {
+                    waitForReindexDone(persistentTask.getId(), listener);
+                } else {
+                    waitForReindexTask(persistentTask.getId(), listener);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                // TODO: This probably should not happen as we are generating the UUID ourselves?
+                if (e instanceof ResourceAlreadyExistsException) {
+                    e = new ElasticsearchStatusException("Cannot create job [" + generatedId +
+                        "] because it has already been created (task exists)", RestStatus.CONFLICT, e);
+                }
+                listener.onFailure(e);
+            }
+        });
+
+    }
+
+    private void waitForReindexDone(String taskId, ActionListener<StartReindexJobAction.Response> listener) {
+        // TODO: Configurable timeout? This currently has a low timeout to prevent tests from hanging.
```

Comment:
can you explain why tests are hanging by indefinitely waiting on completion here?

Reply:
This was to aid with POC development. I removed the timeout.